### PR TITLE
enh(resource-access): handle dynamicaly the datasetfilters

### PR DESCRIPTION
**Pull request by adr-mo**
🏷️ MON-33852

ℹ️ This PR is the next step of the AddRule UseCase. It intends to make filters dynamic, for instance if bam is installed then Business View filter type is available. A lot of refactoring has been done also to validate dataset filters and their hierarchy.
It also fixes an issue regarding linked contacts and contactgroups that were both mandatory. Now if at least one of both is filled then it is OK.

An other PR is available to handle BAM filter types.

- 🟢 PHPStan Level 9 on src/Core/ResourceAccess
- 🟢 PHPCsFixer  
- 🟢 Unit tests were adapted and are all

image

**Fixes** # (issue)

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [ ] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 22.04.x
- [ ] 22.10.x
- [ ] 23.04.x
- [ ] 23.10.x
- [ ] 24.04.x (master)

<h2> How this pull request can be tested ? </h2>

Please describe the **procedure** to verify that the goal of the PR is matched. Provide clear instructions so that it can be **correctly tested**.

Any **relevant details** of the configuration to perform the test should be added.

#### Community contributors & Centreon team

- [x] I have followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: centreon/src/Core/ResourceAccess/Infrastructure/Repository/Providers/HostCategoryRepositoryProvider.php [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        foreach ($resourceIds as $index => $hostCategoryId) {
+            $bindValues[":host_category_id_{$index}"] = $hostCategoryId;
+            $subValues[] = "(:host_category_id_{$index}, :datasetId)";
+        }
+
+        $subQueries = implode(', ', $subValues);
+
+        $request = $this->translateDbName(
+            <<<SQL
+                    INSERT INTO `:db`.acl_resources_hc_relations (hc_id, acl_res_id)
```
Comment (a-launois) [inline centreon/src/Core/ResourceAccess/Infrastructure/Repository/Providers/HostCategoryRepositoryProvider.php:96]: ⚠️ POSSIBLY UNWANTED SIDE SIDE EFFECT
categories and severities ACLs work as "filter", with means that by default you have access to all the categories/severities (aka. when the table acl_resources_hc_relations is empty for a given acl_res_id)
By adding an element in the table, you are effectively **cutting access** to all other categories/severities

File: centreon/src/Core/ResourceAccess/Infrastructure/Repository/Providers/ServiceCategoryRepositoryProvider.php [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        foreach ($resourceIds as $index => $serviceCategoryId) {
+            $bindValues[":service_category_id_{$index}"] = $serviceCategoryId;
+            $subValues[] = "(:service_category_id_{$index}, :datasetId)";
+        }
+
+        $subQueries = implode(', ', $subValues);
+
+        $request = $this->translateDbName(
+            <<<SQL
+                    INSERT INTO `:db`.acl_resources_sc_relations (sc_id, acl_res_id)
```
Comment (a-launois) [inline centreon/src/Core/ResourceAccess/Infrastructure/Repository/Providers/ServiceCategoryRepositoryProvider.php:96]: Same as for host categories:
⚠️ POSSIBLY UNWANTED SIDE SIDE EFFECT
categories and severities ACLs work as "filter", with means that by default you have access to all the categories/severities (aka. when the table is empty for a given acl_res_id)
By adding an element in the table, you are effectively cutting access to all other categories/severities

File: centreon/src/Core/ResourceAccess/Infrastructure/Repository/Providers/HostCategoryRepositoryProvider.php (hunk 1 above)
Comment (adr-mo) [reply centreon/src/Core/ResourceAccess/Infrastructure/Repository/Providers/HostCategoryRepositoryProvider.php:96]: That is indeed the goal. However in Resource Access Management there will not be such thing as exclude / filter resources

Comment (adr-mo) [reply centreon/src/Core/ResourceAccess/Infrastructure/Repository/Providers/HostCategoryRepositoryProvider.php:96]: But I see what you mean

File: centreon/config/packages/Centreon.yaml [hunk 3]
```diff
@@ -307,6 +307,10 @@ services:
             tags: ['configuration.notification.resource.repositories']
         Core\Notification\Application\Repository\NotifiableResourceRequestProviderInterface:
             tags: ['configuration.notification.notifiable.resource.request.provider']
+        Core\ResourceAccess\Domain\Model\DatasetFilter\DatasetFilterTypeInterface:
+            tags: ['resource.access.dataset.filter.types']
+        Core\ResourceAccess\Application\Repository\DatasetRepositoryProviderInterface:
+            tags: ['resource.access.dataset.filter.resource.repository.providers']
```
Comment (callapa) [inline centreon/config/packages/Centreon.yaml:313]: Can you move all framework configurations to a separate file: src/Core/ResourceAccess/Infrastructure/Configuration/symfony.yaml

Don't forget to add the 'src/Core/ResourceAccess' folder to the folder exclusion rule. (see line 38)

File: centreon/www/install/php/Update-24.04.0.php [hunk 4]
```diff
@@ -163,7 +163,7 @@
         CREATE TABLE `dataset_filters` (
             `id` int(11) NOT NULL AUTO_INCREMENT,
             `parent_id` int(11) DEFAULT NULL,
-            `type` enum('host', 'hostgroup', 'host_category', 'servicegroup', 'service_category', 'meta_service', 'service') DEFAULT NULL,
+            `type` enum('host', 'hostgroup', 'host_category', 'servicegroup', 'service_category', 'meta_service', 'service', 'business_view') DEFAULT NULL,
```
Comment (callapa) [inline centreon/www/install/php/Update-24.04.0.php:166]: I think the "business_view" column should be managed directly by the module.

**Review by callapa — APPROVED** [review]

**Review by a-launois — APPROVED** [review]

**Pull request merged**